lighthouse-core/gather/gather-runner.js

```diff
@@ -187,17 +206,56 @@ class GatherRunner {
       const status = `Retrieving: ${gatherer.name}`;
       return chain.then(_ => {
         log.log('status', status);
-        return gatherer.afterPass(options, passData);
-      }).then(ret => {
+        const artifactPromise = Promise.resolve().then(_ => gatherer.afterPass(options, passData));
+        gathererResults[gatherer.name].push(artifactPromise);
+        return artifactPromise.catch(err => {
+          if (!err.recoverable) {
+            throw err;
+          }
+        });
+      }).then(_ => {
         log.verbose('statusEnd', status);
-        return ret;
       });
     }, pass);
 
     // Resolve on tracing data using passName from config.
     return pass.then(_ => passData);
   }
 
+  /**
+   * Takes the results of each gatherer phase for each gatherer and uses the
+   * last produced value (that's not undefined) as the artifact for that
+   * gatherer. If a recoverable error was rejected from a gatherer phase,
+   * uses that error as the artifact instead.
+   * @param {!Object<!Array<!Promise<*>>} gathererResults
+   * @return {!Promise<!Artifacts>}
+   */
+  static collectArtifacts(gathererResults) {
+    const artifacts = {};
+
+    return Object.keys(gathererResults).reduce((chain, gathererName) => {
+      return chain.then(_ => {
+        const phaseResultsPromises = gathererResults[gathererName];
+        return Promise.all(phaseResultsPromises).then(phaseResults => {
+          // Take last defined pass result as artifact.
+          const artifact = phaseResults.reduceRight((prev, curr) => {
```

Comment:
`reduceRight` makes this a bit confusing to understand for how simple a job it's doing. what about just a filter and taking the last one?